File: Material/TransitSoft-DobleBD/TransitSoftDBManager/DBManager.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TransitSoft.Db.Util;

namespace TransitSoft.Db
{
    public abstract class DBManager
    {
        protected DbConnection conexion;
        protected string baseDeDatos;
        protected string nombreDeHost;
        protected string puerto;
        protected string usuario;
        protected string contraseña;
        protected static DBManager dbManager;

        // Nueva instancia para la segunda base de datos
        protected static DBManager dbManager2;

        protected DBManager()
        {
            //constructor protegido para evitar que se creen instancias.
            //Solo se podrá crear una instancia y esta debe hacerse usando el 
            //método getInstance()
        }

        // Instancia principal (mantiene compatibilidad con código existente)
        public static DBManager Instance
        {
            get
            {
                if (dbManager == null)
                {
                    DBManager.CreateInstance();
                }
                return dbManager;
            }
        }

        // Nueva propiedad para la segunda base de datos
        public static DBManager Instance2
        {
            get
            {
                if (dbManager2 == null)
                {
                    DBManager.CreateInstance2();
                }
                return dbManager2;
            }
        }

        private static void CreateInstance()
        {
            if (DBManager.dbManager == null)
            {
                if (DBManager.ObtenerMotorDeBaseDeDatos() == MotorDeBaseDeDatos.MYSQL)
                    DBManagerMySQL.CreateConcreteInstance();
                else
                    DBManagerMSSQL.CreateConcreteInstance();
                DBManager.dbManager.LeerArchivoDeConfiguracion(""); // BD principal
            }
        }

        // Nuevo método para crear la segunda instancia
        private static void CreateInstance2()
        {
            if (DBManager.dbManager2 == null)
            {
                if (DBManager.ObtenerMotorDeBaseDeDatos2() == MotorDeBaseDeDatos.MYSQL)
                    DBManagerMySQL.CreateConcreteInstance2();
                else
                    DBManagerMSSQL.CreateConcreteInstance2();
                DBManager.dbManager2.LeerArchivoDeConfiguracion("2"); // BD secundaria
            }
        }

        private static MotorDeBaseDeDatos ObtenerMotorDeBaseDeDatos()
        {
            string tipoDeBaseDeDatos = ConfigurationManager.AppSettings["tipoDeBaseDeDatos"];
            if (tipoDeBaseDeDatos.Equals(MotorDeBaseDeDatos.MYSQL.ToString()))
                return MotorDeBaseDeDatos.MYSQL;
            else
                return MotorDeBaseDeDatos.MSSQL;
        }

        // Nuevo método para obtener el motor de la segunda BD
        private static MotorDeBaseDeDatos ObtenerMotorDeBaseDeDatos2()
        {
            string tipoDeBaseDeDatos = ConfigurationManager.AppSettings["tipoDeBaseDeDatos2"];
            if (string.IsNullOrEmpty(tipoDeBaseDeDatos))
            {
                // Si no está configurada, usar el mismo tipo que la primera
                return ObtenerMotorDeBaseDeDatos();
            }
            if (tipoDeBaseDeDatos.Equals(MotorDeBaseDeDatos.MYSQL.ToString()))
                return MotorDeBaseDeDatos.MYSQL;
            else
                return MotorDeBaseDeDatos.MSSQL;
        }

        // Modificado para aceptar un sufijo
        private void LeerArchivoDeConfiguracion(string sufijo)
        {
            this.baseDeDatos = ConfigurationManager.AppSettings[$"baseDeDatos{sufijo}"];
            this.nombreDeHost = ConfigurationManager.AppSettings[$"nombreDeHost{sufijo}"];
            this.puerto = ConfigurationManager.AppSettings[$"puerto{sufijo}"];
            this.usuario = ConfigurationManager.AppSettings[$"usuario{sufijo}"];
            this.contraseña = ConfigurationManager.AppSettings[$"contraseña{sufijo}"];
        }

        public DbConnection Connection
        {
            get
            {
                string cadenaDeConexion = this.ObtenerCadenaDeConexion();
                this.RealizarConexionABaseDeDatos(cadenaDeConexion);
                return this.conexion;
            }
        }

        protected abstract void RealizarConexionABaseDeDatos(string cadenaDeConexion);

        protected abstract string ObtenerCadenaDeConexion();

        public abstract DbCommand CrearComando();

        public abstract string RetornarSQLParaUltimoAutoGenerado();

        public static string Cifrar(string texto)
        {
            return Convert.ToBase64String(Encoding.Unicode.GetBytes(texto));
        }

        public static string Descifrar(string textoEncriptado)
        {
            return Encoding.Unicode.GetString(Convert.FromBase64String(textoEncriptado));
        }
    }
}

File: EX2/EX2/Material/TransitSoft_ConexionMultiple-SOAP/TransitSoftDBManager/src/main/java/pe/edu/pucp/transitsoft/db/DBManager.java
package pe.edu.pucp.transitsoft.db;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;
import java.util.logging.Level;
import java.util.logging.Logger;
import pe.edu.pucp.transitsoft.db.util.Cifrado;
import pe.edu.pucp.transitsoft.db.util.MotorDeBaseDeDatos;

public abstract class DBManager {

    private static final String ARCHIVO_CONFIGURACION = "db.properties";
    
    private Connection conexion;
    private String driver;
    protected String tipo_de_driver;
    protected String base_de_datos;
    protected String nombre_de_host;
    protected String puerto;
    private String usuario;
    private String contraseña;
    
    // Tres instancias estáticas independientes
    private static DBManager dbManager = null;   // BD Principal
    private static DBManager dbManager2 = null;  // BD Secundaria
    private static DBManager dbManager3 = null;  // BD Terciaria

    protected DBManager() {
        //constructor protegido para evitar que se creen instancias.
        //Solo se podrá crear una instancia y esta debe hacerse usando el 
        //método getInstance()
    }

    // Instancia principal (mantiene compatibilidad)
    public static DBManager getInstance() {
        if (DBManager.dbManager == null) {
            DBManager.createInstance();
        }
        return DBManager.dbManager;
    }

    // Segunda instancia
    public static DBManager getInstance2() {
        if (DBManager.dbManager2 == null) {
            DBManager.createInstance2();
        }
        return DBManager.dbManager2;
    }

    // Tercera instancia
    public static DBManager getInstance3() {
        if (DBManager.dbManager3 == null) {
            DBManager.createInstance3();
        }
        return DBManager.dbManager3;
    }

    private static void createInstance() {
        if (DBManager.dbManager == null) {
            MotorDeBaseDeDatos motor = DBManager.obtenerMotorDeBaseDeDato("");
            DBManager.dbManager = crearInstanciaSegunMotor(motor);
            DBManager.dbManager.leer_archivo_de_propiedades("");
        }
    }

    private static void createInstance2() {
        if (DBManager.dbManager2 == null) {
            MotorDeBaseDeDatos motor = DBManager.obtenerMotorDeBaseDeDato("2");
            DBManager.dbManager2 = crearInstanciaSegunMotor(motor);
            DBManager.dbManager2.leer_archivo_de_propiedades("2");
        }
    }

    private static void createInstance3() {
        if (DBManager.dbManager3 == null) {
            MotorDeBaseDeDatos motor = DBManager.obtenerMotorDeBaseDeDato("3");
            DBManager.dbManager3 = crearInstanciaSegunMotor(motor);
            DBManager.dbManager3.leer_archivo_de_propiedades("3");
        }
    }

    private static DBManager crearInstanciaSegunMotor(MotorDeBaseDeDatos motor) {
        if (motor == MotorDeBaseDeDatos.MYSQL) {
            return new DBManagerMySQL();
        } else if (motor == MotorDeBaseDeDatos.MSSQL) {
            return new DBManagerMSSQL();
        } else if (motor == MotorDeBaseDeDatos.ORACLE) {
            return new DBManagerOracle();
        }
        return null;
    }

    public Connection getConnection() {
        try {
            Class.forName(this.driver);
            this.conexion = DriverManager.getConnection(
                getURL(), 
                this.usuario, 
                Cifrado.descifrarMD5(this.contraseña)
            );
        } catch (ClassNotFoundException ex) {
            System.err.println("Error al generar la conexión - " + ex);
        } catch (SQLException ex) {
            Logger.getLogger(DBManager.class.getName()).log(Level.SEVERE, null, ex);
        }
        return conexion;
    }

    protected abstract String getURL();

    private void leer_archivo_de_propiedades(String sufijo) {
        Properties properties = new Properties();
        try {
            String nmArchivoConf = "/" + ARCHIVO_CONFIGURACION;

            properties.load(this.getClass().getResourceAsStream(nmArchivoConf));
            this.driver = properties.getProperty("driver" + sufijo);
            this.tipo_de_driver = properties.getProperty("tipo_de_driver" + sufijo);
            this.base_de_datos = properties.getProperty("base_de_datos" + sufijo);
            this.nombre_de_host = properties.getProperty("nombre_de_host" + sufijo);
            this.puerto = properties.getProperty("puerto" + sufijo);
            this.usuario = properties.getProperty("usuario" + sufijo);
            this.contraseña = properties.getProperty("contrasenha" + sufijo);
        } catch (FileNotFoundException ex) {
            System.err.println("Error al leer el archivo de propiedades - " + ex);
        } catch (IOException ex) {
            System.err.println("Error al leer el archivo de propiedades - " + ex);
        }
    }
    
    private static MotorDeBaseDeDatos obtenerMotorDeBaseDeDato(String sufijo) {
        Properties properties = new Properties();
        try {
            String nmArchivoConf = "/" + ARCHIVO_CONFIGURACION;
            properties.load(DBManager.class.getResourceAsStream(nmArchivoConf));            
            String tipo_de_driver = properties.getProperty("tipo_de_driver" + sufijo);
            
            if (tipo_de_driver == null) {
                System.err.println("⚠ No se encontró tipo_de_driver" + sufijo + ", usando MySQL por defecto");
                return MotorDeBaseDeDatos.MYSQL;
            }
            
            // Verificar el tipo de driver por su prefijo
            if (tipo_de_driver.startsWith("jdbc:mysql")) {
                return MotorDeBaseDeDatos.MYSQL;
            } else if (tipo_de_driver.startsWith("jdbc:sqlserver")) {
                return MotorDeBaseDeDatos.MSSQL;
            } else if (tipo_de_driver.startsWith("jdbc:oracle")) {
                return MotorDeBaseDeDatos.ORACLE;
            }
            
            // Si no coincide con ninguno, intentar por nombre del driver
            String driver = properties.getProperty("driver" + sufijo);
            if (driver != null) {
                if (driver.contains("mysql")) {
                    return MotorDeBaseDeDatos.MYSQL;
                } else if (driver.contains("sqlserver")) {
                    return MotorDeBaseDeDatos.MSSQL;
                } else if (driver.contains("oracle")) {
                    return MotorDeBaseDeDatos.ORACLE;
                }
            }
            
            // Fallback a MySQL por defecto
            System.err.println("⚠ No se pudo determinar el motor de BD" + sufijo + ", usando MySQL por defecto");
            return MotorDeBaseDeDatos.MYSQL;
            
        } catch (FileNotFoundException ex) {
            System.err.println("Error al leer el archivo de propiedades - " + ex);
        } catch (IOException ex) {
            System.err.println("Error al leer el archivo de propiedades - " + ex);
        }
        return MotorDeBaseDeDatos.MYSQL;
    }
    
    public abstract String retornarSQLParaUltimoAutoGenerado();
}